distributed/comm/ucx.py

```diff
@@ -169,19 +196,44 @@ async def write(
                 frames = await to_frames(
                     msg, serializers=serializers, on_error=on_error
                 )
-                send_frames = [
-                    each_frame for each_frame in frames if len(each_frame) > 0
-                ]
+                nframes = len(frames)
+                cuda_frames = tuple(
+                    hasattr(f, "__cuda_array_interface__") for f in frames
+                )
+                sizes = tuple(nbytes(f) for f in frames)
+                host_frames = host_array(
+                    sum(
+                        each_size
+                        for is_cuda, each_size in zip(cuda_frames, sizes)
+                        if not is_cuda
+                    )
+                )
+                device_frames = device_array(
+                    sum(
+                        each_size
+                        for is_cuda, each_size in zip(cuda_frames, sizes)
+                        if is_cuda
+                    )
+                )
```

Comment:
We're looping over `frames` or some computed attribute of frames four times here in separate loops. Would it make sense to fuse these into a single loop?